### PR TITLE
JIT: relax immediate dominator check for jump threading

### LINKED ISSUE #53501: Jump threading regression in preview 4
### Description
.NET 6 preview 1 introduced jump threading, I'm trying it with .NET 6 preview 4 with the case mentioned in PR #46257.

```csharp
static int Y(object o)
{
    int result = 0;
    if (o.GetType() == typeof(string)) result++;
    if (o.GetType() == typeof(string)) result++;
    if (o.GetType() == typeof(string)) result++;
    if (o.GetType() == typeof(string)) result++;
    return result;	
}
```

But got following result:

```asm
L0000 xor eax, eax
L0002 mov rdx, [rcx]
L0005 mov rcx, 0x7ffb84ecc1e8
L000f cmp rdx, rcx
L0012 jne short L001b ; <-- supposed to be jne short L002e
L0014 mov eax, 1
L0019 inc eax
L001b mov rcx, 0x7ffb84ecc1e8 ; | <-- redundant check
L0025 cmp rdx, rcx            ; |
L0028 jne short L002e         ;_|
L002a inc eax
L002c inc eax
L002e ret
```

which is supposed to be:
```asm
       xor      eax, eax
       mov      rdx, 0xD1FFAB1E
       cmp      qword ptr [rcx], rdx
       jne      SHORT G_M10236_IG03
       mov      eax, 1
       inc      eax
       inc      eax
       inc      eax
G_M10236_IG03
       ret
```

Is this a regression?

### Configuration

.NET 6 Preview 4

### Regression?

Yes

### PR DESCRIPTION
**Pull request by AndyAyersMS**
When jump threading, we had been insisting that the redundant branch be
the immediate dominator (idom) of the branch being optimized. But now that we
are doing exact reachability checks, we can consider threading based on
redundant branches higher in the (original) dominator tree, provided that the
in-between branches have also been optimized.

This situation arises when there are repeated redundant branches in if-then
or if-then-else chains. Say there are 3 such branches. The algorithm works
top-down in the original dominator tree. To begin with, the second branch
in the chain is jump-threaded since it redundant to the first. As part
of this the second branch's block is modified to fall through. Now the
third branch's idom is the block that held the second branch, but this block
no longer ends with a branch. So the nearest redundant dominating branch is the
first branch in the chain. And this arrangement blocks jump-threading the third
branch as the redundant branch block is not in the idom block.

Jump threading will work corectly so long as there are unique paths from the
redundant branch outcomes to the branch being optimized. So we have appropriate
safety checks and can consider threading based on higher dominating branches.

Resolves #53501.

If we have a repeated chain of redundant branches

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
cc @dotnet/jit-contrib 

Test case diff
```diff
 ;  V00 arg0         [V00,T01] (  3,  3   )     ref  ->  rcx         class-hnd
 ;  V01 loc0         [V01,T00] (  9,  9   )     int  ->  rax
 ;# V02 OutArgs      [V02    ] (  1,  1   )  lclBlk ( 0) [rsp+00H]   "OutgoingArgSpace"
-;  V03 cse0         [V03,T02] (  3,  3   )    long  ->  rdx         "CSE - aggressive"
 ;
 ; Lcl frame size = 0

@@ -17,25 +16,19 @@ G_M10236_IG01:
                                                ;; bbWeight=1    PerfScore 0.00
 G_M10236_IG02:
        xor      eax, eax
-       mov      rdx, qword ptr [rcx]
-       mov      rcx, 0xD1FFAB1E
-       cmp      rdx, rcx
+       mov      rdx, 0xD1FFAB1E
+       cmp      qword ptr [rcx], rdx
        jne      SHORT G_M10236_IG03
        mov      eax, 1
        inc      eax
-                                               ;; bbWeight=1    PerfScore 4.25
-G_M10236_IG03:
-       mov      rcx, 0xD1FFAB1E
-       cmp      rdx, rcx
-       jne      SHORT G_M10236_IG04
        inc      eax
        inc      eax
-                                               ;; bbWeight=1    PerfScore 2.00
-G_M10236_IG04:
+                                               ;; bbWeight=1    PerfScore 4.50
+G_M10236_IG03:
        ret
                                                ;; bbWeight=1    PerfScore 1.00

-; Total bytes of code 47, prolog size 0, PerfScore 11.95, instruction count 13, allocated bytes for code 47 (MethodHash=d213d803) for method X:Y(Object):int
+; Total bytes of code 29, prolog size 0, PerfScore 8.40, instruction count 9, allocated bytes for code 29 (MethodHash=d213d803) for method X:Y(Object):int
```

SPMI diffs

**asm.benchmarks.run.windows.x64.checked** 
```

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 8120
Total bytes of diff: 8097
Total bytes of delta: -23 (-0.28% of base)
    diff is an improvement.
```
<details>

<summary>Detail diffs</summary>

```


Top file improvements (bytes):
         -16 : 16257.dasm (-0.42% of base)
          -7 : 15717.dasm (-0.17% of base)

2 total files with Code Size differences (2 improved, 0 regressed), 1 unchanged.

Top method improvements (bytes):
         -16 (-0.42% of base) : 16257.dasm - ILStubClass:IL_STUB_StructMarshal(byref,long,int,byref)
          -7 (-0.17% of base) : 15717.dasm - ILStubClass:IL_STUB_StructMarshal(byref,long,int,byref)

Top method improvements (percentages):
         -16 (-0.42% of base) : 16257.dasm - ILStubClass:IL_STUB_StructMarshal(byref,long,int,byref)
          -7 (-0.17% of base) : 15717.dasm - ILStubClass:IL_STUB_StructMarshal(byref,long,int,byref)

2 total methods with Code Size differences (2 improved, 0 regressed), 1 unchanged.

```

</details>

--------------------------------------------------------------------------------

**asm.coreclr_tests.pmi.windows.x64.checked** 
```

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 91960
Total bytes of diff: 90047
Total bytes of delta: -1913 (-2.08% of base)
    diff is an improvement.
```
<details>

<summary>Detail diffs</summary>

```


Top file regressions (bytes):
           9 : 88678.dasm (2.74% of base)
           9 : 88596.dasm (2.74% of base)
           8 : 136620.dasm (2.99% of base)
           8 : 87358.dasm (3.05% of base)
           8 : 89861.dasm (2.91% of base)
           8 : 87360.dasm (2.52% of base)
           8 : 136702.dasm (2.75% of base)
           8 : 89832.dasm (2.91% of base)
           8 : 136622.dasm (2.45% of base)
           8 : 89837.dasm (3.05% of base)
           8 : 89839.dasm (2.52% of base)
           8 : 88636.dasm (2.91% of base)
           6 : 136660.dasm (2.03% of base)
           6 : 87440.dasm (2.11% of base)
           6 : 88363.dasm (2.08% of base)
           6 : 88775.dasm (2.08% of base)
           6 : 87398.dasm (2.08% of base)

Top file improvements (bytes):
        -125 : 213372.dasm (-6.66% of base)
         -75 : 87841.dasm (-25.86% of base)
         -69 : 87432.dasm (-24.30% of base)
         -69 : 88929.dasm (-24.30% of base)
         -69 : 88519.dasm (-24.30% of base)
         -67 : 89079.dasm (-25.97% of base)
         -63 : 89082.dasm (-28.51% of base)
         -61 : 88671.dasm (-24.21% of base)
         -61 : 88631.dasm (-24.21% of base)
         -61 : 88442.dasm (-24.21% of base)
         -61 : 89039.dasm (-24.21% of base)
         -61 : 88854.dasm (-24.21% of base)
         -57 : 88674.dasm (-26.51% of base)
         -35 : 89097.dasm (-18.92% of base)
         -31 : 241946.dasm (-2.87% of base)
         -29 : 88605.dasm (-16.20% of base)
         -29 : 88828.dasm (-16.20% of base)
         -29 : 88689.dasm (-16.20% of base)
         -29 : 88416.dasm (-16.20% of base)
         -29 : 89013.dasm (-16.20% of base)

213 total files with Code Size differences (196 improved, 17 regressed), 8 unchanged.

Top method regressions (bytes):
           9 ( 2.74% of base) : 88678.dasm - testout1:Sub_Funclet_49():int
           9 ( 2.74% of base) : 88596.dasm - testout1:Sub_Funclet_145():int
           8 ( 2.99% of base) : 136620.dasm - testout1:Sub_Funclet_141():this
           8 ( 3.05% of base) : 87358.dasm - testout1:Sub_Funclet_141():int
           8 ( 2.91% of base) : 89861.dasm - testout1:Sub_Funclet_9():int
           8 ( 2.52% of base) : 87360.dasm - testout1:Sub_Funclet_143():int
           8 ( 2.75% of base) : 136702.dasm - testout1:Sub_Funclet_11():this
           8 ( 2.91% of base) : 89832.dasm - testout1:Sub_Funclet_121():int
           8 ( 2.45% of base) : 136622.dasm - testout1:Sub_Funclet_143():this
           8 ( 3.05% of base) : 89837.dasm - testout1:Sub_Funclet_77():int
           8 ( 2.52% of base) : 89839.dasm - testout1:Sub_Funclet_79():int
           8 ( 2.91% of base) : 88636.dasm - testout1:Sub_Funclet_121():int
           6 ( 2.03% of base) : 136660.dasm - testout1:Sub_Funclet_117():this
           6 ( 2.11% of base) : 87440.dasm - testout1:Sub_Funclet_11():int
           6 ( 2.08% of base) : 88363.dasm - testout1:Sub_Funclet_357():int
           6 ( 2.08% of base) : 88775.dasm - testout1:Sub_Funclet_357():int
           6 ( 2.08% of base) : 87398.dasm - testout1:Sub_Funclet_117():int

Top method improvements (bytes):
        -125 (-6.66% of base) : 213372.dasm - testout1:Sub_Funclet_0():int
         -75 (-25.86% of base) : 87841.dasm - testout1:Sub_Funclet_1():int
         -69 (-24.30% of base) : 87432.dasm - testout1:Sub_Funclet_1():int
         -69 (-24.30% of base) : 88929.dasm - testout1:Sub_Funclet_193():int
         -69 (-24.30% of base) : 88519.dasm - testout1:Sub_Funclet_193():int
         -67 (-25.97% of base) : 89079.dasm - testout1:Sub_Funclet_36():int
         -63 (-28.51% of base) : 89082.dasm - testout1:Sub_Funclet_39():int
         -61 (-24.21% of base) : 88671.dasm - testout1:Sub_Funclet_36():int
         -61 (-24.21% of base) : 88631.dasm - testout1:Sub_Funclet_116():int
         -61 (-24.21% of base) : 88442.dasm - testout1:Sub_Funclet_308():int
         -61 (-24.21% of base) : 89039.dasm - testout1:Sub_Funclet_116():int
         -61 (-24.21% of base) : 88854.dasm - testout1:Sub_Funclet_308():int
         -57 (-26.51% of base) : 88674.dasm - testout1:Sub_Funclet_39():int
         -35 (-18.92% of base) : 89097.dasm - testout1:Sub_Funclet_10():int
         -31 (-2.87% of base) : 241946.dasm - CseTest.Test_Main:Main():int
         -29 (-16.20% of base) : 88605.dasm - testout1:Sub_Funclet_154():int
         -29 (-16.20% of base) : 88828.dasm - testout1:Sub_Funclet_346():int
         -29 (-16.20% of base) : 88689.dasm - testout1:Sub_Funclet_10():int
         -29 (-16.20% of base) : 88416.dasm - testout1:Sub_Funclet_346():int
         -29 (-16.20% of base) : 89013.dasm - testout1:Sub_Funclet_154():int

Top method regressions (percentages):
           8 ( 3.05% of base) : 87358.dasm - testout1:Sub_Funclet_141():int
           8 ( 3.05% of base) : 89837.dasm - testout1:Sub_Funclet_77():int
           8 ( 2.99% of base) : 136620.dasm - testout1:Sub_Funclet_141():this
           8 ( 2.91% of base) : 89861.dasm - testout1:Sub_Funclet_9():int
           8 ( 2.91% of base) : 89832.dasm - testout1:Sub_Funclet_121():int
           8 ( 2.91% of base) : 88636.dasm - testout1:Sub_Funclet_121():int
           8 ( 2.75% of base) : 136702.dasm - testout1:Sub_Funclet_11():this
           9 ( 2.74% of base) : 88678.dasm - testout1:Sub_Funclet_49():int
           9 ( 2.74% of base) : 88596.dasm - testout1:Sub_Funclet_145():int
           8 ( 2.52% of base) : 87360.dasm - testout1:Sub_Funclet_143():int
           8 ( 2.52% of base) : 89839.dasm - testout1:Sub_Funclet_79():int
           8 ( 2.45% of base) : 136622.dasm - testout1:Sub_Funclet_143():this
           6 ( 2.11% of base) : 87440.dasm - testout1:Sub_Funclet_11():int
           6 ( 2.08% of base) : 88363.dasm - testout1:Sub_Funclet_357():int
           6 ( 2.08% of base) : 88775.dasm - testout1:Sub_Funclet_357():int
           6 ( 2.08% of base) : 87398.dasm - testout1:Sub_Funclet_117():int
           6 ( 2.03% of base) : 136660.dasm - testout1:Sub_Funclet_117():this

Top method improvements (percentages):
         -63 (-28.51% of base) : 89082.dasm - testout1:Sub_Funclet_39():int
         -57 (-26.51% of base) : 88674.dasm - testout1:Sub_Funclet_39():int
         -67 (-25.97% of base) : 89079.dasm - testout1:Sub_Funclet_36():int
         -75 (-25.86% of base) : 87841.dasm - testout1:Sub_Funclet_1():int
         -69 (-24.30% of base) : 87432.dasm - testout1:Sub_Funclet_1():int
         -69 (-24.30% of base) : 88929.dasm - testout1:Sub_Funclet_193():int
         -69 (-24.30% of base) : 88519.dasm - testout1:Sub_Funclet_193():int
         -61 (-24.21% of base) : 88671.dasm - testout1:Sub_Funclet_36():int
         -61 (-24.21% of base) : 88631.dasm - testout1:Sub_Funclet_116():int
         -61 (-24.21% of base) : 88442.dasm - testout1:Sub_Funclet_308():int
         -61 (-24.21% of base) : 89039.dasm - testout1:Sub_Funclet_116():int
         -61 (-24.21% of base) : 88854.dasm - testout1:Sub_Funclet_308():int
         -35 (-18.92% of base) : 89097.dasm - testout1:Sub_Funclet_10():int
         -29 (-16.20% of base) : 88605.dasm - testout1:Sub_Funclet_154():int
         -29 (-16.20% of base) : 88828.dasm - testout1:Sub_Funclet_346():int
         -29 (-16.20% of base) : 88689.dasm - testout1:Sub_Funclet_10():int
         -29 (-16.20% of base) : 88416.dasm - testout1:Sub_Funclet_346():int
         -29 (-16.20% of base) : 89013.dasm - testout1:Sub_Funclet_154():int
         -20 (-6.99% of base) : 136694.dasm - testout1:Sub_Funclet_1():this
         -20 (-6.67% of base) : 89060.dasm - testout1:Sub_Funclet_78():int

213 total methods with Code Size differences (196 improved, 17 regressed), 8 unchanged.

```

</details>

--------------------------------------------------------------------------------

**asm.libraries.crossgen.windows.x64.checked** 
```

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 6891
Total bytes of diff: 6779
Total bytes of delta: -112 (-1.63% of base)
    diff is an improvement.
```
<details>

<summary>Detail diffs</summary>

```


Top file improvements (bytes):
         -42 : 108700.dasm (-13.46% of base)
         -29 : 37435.dasm (-3.18% of base)
         -25 : 33681.dasm (-1.57% of base)
          -6 : 44892.dasm (-2.59% of base)
          -5 : 136946.dasm (-0.63% of base)
          -5 : 136867.dasm (-0.63% of base)

6 total files with Code Size differences (6 improved, 0 regressed), 3 unchanged.

Top method improvements (bytes):
         -42 (-13.46% of base) : 108700.dasm - Microsoft.CSharp.RuntimeBinder.Semantics.SymbolLoader:HasArrayConversionToInterface(Microsoft.CSharp.RuntimeBinder.Semantics.ArrayType,Microsoft.CSharp.RuntimeBinder.Semantics.CType):bool
         -29 (-3.18% of base) : 37435.dasm - Microsoft.CodeAnalysis.CSharp.Symbols.SourcePropertySymbol:CheckModifiers(Microsoft.CodeAnalysis.Location,bool,Microsoft.CodeAnalysis.DiagnosticBag):this
         -25 (-1.57% of base) : 33681.dasm - Microsoft.CodeAnalysis.CSharp.Symbols.SourceEventSymbol:CheckModifiersAndType(Microsoft.CodeAnalysis.DiagnosticBag):this
          -6 (-2.59% of base) : 44892.dasm - Microsoft.CodeAnalysis.VisualBasic.LambdaUtilities:IsNonUserCodeQueryLambda(Microsoft.CodeAnalysis.SyntaxNode):bool
          -5 (-0.63% of base) : 136946.dasm - System.Drawing.Drawing2D.PathGradientBrush:set_Blend(System.Drawing.Drawing2D.Blend):this
          -5 (-0.63% of base) : 136867.dasm - System.Drawing.Drawing2D.LinearGradientBrush:set_Blend(System.Drawing.Drawing2D.Blend):this

Top method improvements (percentages):
         -42 (-13.46% of base) : 108700.dasm - Microsoft.CSharp.RuntimeBinder.Semantics.SymbolLoader:HasArrayConversionToInterface(Microsoft.CSharp.RuntimeBinder.Semantics.ArrayType,Microsoft.CSharp.RuntimeBinder.Semantics.CType):bool
         -29 (-3.18% of base) : 37435.dasm - Microsoft.CodeAnalysis.CSharp.Symbols.SourcePropertySymbol:CheckModifiers(Microsoft.CodeAnalysis.Location,bool,Microsoft.CodeAnalysis.DiagnosticBag):this
          -6 (-2.59% of base) : 44892.dasm - Microsoft.CodeAnalysis.VisualBasic.LambdaUtilities:IsNonUserCodeQueryLambda(Microsoft.CodeAnalysis.SyntaxNode):bool
         -25 (-1.57% of base) : 33681.dasm - Microsoft.CodeAnalysis.CSharp.Symbols.SourceEventSymbol:CheckModifiersAndType(Microsoft.CodeAnalysis.DiagnosticBag):this
          -5 (-0.63% of base) : 136946.dasm - System.Drawing.Drawing2D.PathGradientBrush:set_Blend(System.Drawing.Drawing2D.Blend):this
          -5 (-0.63% of base) : 136867.dasm - System.Drawing.Drawing2D.LinearGradientBrush:set_Blend(System.Drawing.Drawing2D.Blend):this

6 total methods with Code Size differences (6 improved, 0 regressed), 3 unchanged.

```

</details>

--------------------------------------------------------------------------------

**asm.libraries.crossgen2.windows.x64.checked** 
```

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 4340
Total bytes of diff: 4282
Total bytes of delta: -58 (-1.34% of base)
    diff is an improvement.
```
<details>

<summary>Detail diffs</summary>

```


Top file improvements (bytes):
         -42 : 210691.dasm (-13.42% of base)
          -6 : 82975.dasm (-2.59% of base)
          -5 : 174929.dasm (-0.65% of base)
          -5 : 174851.dasm (-0.65% of base)

4 total files with Code Size differences (4 improved, 0 regressed), 3 unchanged.

Top method improvements (bytes):
         -42 (-13.42% of base) : 210691.dasm - Microsoft.CSharp.RuntimeBinder.Semantics.SymbolLoader:HasArrayConversionToInterface(Microsoft.CSharp.RuntimeBinder.Semantics.ArrayType,Microsoft.CSharp.RuntimeBinder.Semantics.CType):bool
          -6 (-2.59% of base) : 82975.dasm - Microsoft.CodeAnalysis.VisualBasic.LambdaUtilities:IsNonUserCodeQueryLambda(Microsoft.CodeAnalysis.SyntaxNode):bool
          -5 (-0.65% of base) : 174929.dasm - System.Drawing.Drawing2D.LinearGradientBrush:set_Blend(System.Drawing.Drawing2D.Blend):this
          -5 (-0.65% of base) : 174851.dasm - System.Drawing.Drawing2D.PathGradientBrush:set_Blend(System.Drawing.Drawing2D.Blend):this

Top method improvements (percentages):
         -42 (-13.42% of base) : 210691.dasm - Microsoft.CSharp.RuntimeBinder.Semantics.SymbolLoader:HasArrayConversionToInterface(Microsoft.CSharp.RuntimeBinder.Semantics.ArrayType,Microsoft.CSharp.RuntimeBinder.Semantics.CType):bool
          -6 (-2.59% of base) : 82975.dasm - Microsoft.CodeAnalysis.VisualBasic.LambdaUtilities:IsNonUserCodeQueryLambda(Microsoft.CodeAnalysis.SyntaxNode):bool
          -5 (-0.65% of base) : 174929.dasm - System.Drawing.Drawing2D.LinearGradientBrush:set_Blend(System.Drawing.Drawing2D.Blend):this
          -5 (-0.65% of base) : 174851.dasm - System.Drawing.Drawing2D.PathGradientBrush:set_Blend(System.Drawing.Drawing2D.Blend):this

4 total methods with Code Size differences (4 improved, 0 regressed), 3 unchanged.

```

</details>

--------------------------------------------------------------------------------

**asm.libraries.pmi.windows.x64.checked** 
```

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 5057
Total bytes of diff: 4988
Total bytes of delta: -69 (-1.36% of base)
    diff is an improvement.
```
<details>

<summary>Detail diffs</summary>

```


Top file regressions (bytes):
           9 : 143805.dasm (1.26% of base)

Top file improvements (bytes):
         -42 : 141092.dasm (-13.00% of base)
         -26 : 67485.dasm (-19.40% of base)
          -5 : 175978.dasm (-0.53% of base)
          -5 : 175895.dasm (-0.53% of base)

5 total files with Code Size differences (4 improved, 1 regressed), 1 unchanged.

Top method regressions (bytes):
           9 ( 1.26% of base) : 143805.dasm - Microsoft.Extensions.FileProviders.Physical.Internal.PathUtils:PathNavigatesAboveRoot(System.String):bool

Top method improvements (bytes):
         -42 (-13.00% of base) : 141092.dasm - Microsoft.CSharp.RuntimeBinder.Semantics.SymbolLoader:HasArrayConversionToInterface(Microsoft.CSharp.RuntimeBinder.Semantics.ArrayType,Microsoft.CSharp.RuntimeBinder.Semantics.CType):bool
         -26 (-19.40% of base) : 67485.dasm - Microsoft.CodeAnalysis.VisualBasic.LambdaUtilities:IsNonUserCodeQueryLambda(Microsoft.CodeAnalysis.SyntaxNode):bool
          -5 (-0.53% of base) : 175978.dasm - System.Drawing.Drawing2D.PathGradientBrush:set_Blend(System.Drawing.Drawing2D.Blend):this
          -5 (-0.53% of base) : 175895.dasm - System.Drawing.Drawing2D.LinearGradientBrush:set_Blend(System.Drawing.Drawing2D.Blend):this

Top method regressions (percentages):
           9 ( 1.26% of base) : 143805.dasm - Microsoft.Extensions.FileProviders.Physical.Internal.PathUtils:PathNavigatesAboveRoot(System.String):bool

Top method improvements (percentages):
         -26 (-19.40% of base) : 67485.dasm - Microsoft.CodeAnalysis.VisualBasic.LambdaUtilities:IsNonUserCodeQueryLambda(Microsoft.CodeAnalysis.SyntaxNode):bool
         -42 (-13.00% of base) : 141092.dasm - Microsoft.CSharp.RuntimeBinder.Semantics.SymbolLoader:HasArrayConversionToInterface(Microsoft.CSharp.RuntimeBinder.Semantics.ArrayType,Microsoft.CSharp.RuntimeBinder.Semantics.CType):bool
          -5 (-0.53% of base) : 175978.dasm - System.Drawing.Drawing2D.PathGradientBrush:set_Blend(System.Drawing.Drawing2D.Blend):this
          -5 (-0.53% of base) : 175895.dasm - System.Drawing.Drawing2D.LinearGradientBrush:set_Blend(System.Drawing.Drawing2D.Blend):this

5 total methods with Code Size differences (4 improved, 1 regressed), 1 unchanged.

```

</details>

--------------------------------------------------------------------------------

**asm.libraries_tests.pmi.windows.x64.checked** 
```

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 19199
Total bytes of diff: 18964
Total bytes of delta: -235 (-1.22% of base)
    diff is an improvement.
```
<details>

<summary>Detail diffs</summary>

```


Top file regressions (bytes):
           4 : 127375.dasm (0.60% of base)

Top file improvements (bytes):
        -131 : 126763.dasm (-1.49% of base)
         -43 : 126729.dasm (-25.60% of base)
         -23 : 126559.dasm (-2.85% of base)
         -13 : 126582.dasm (-7.10% of base)
          -5 : 10565.dasm (-0.94% of base)
          -5 : 212745.dasm (-0.47% of base)
          -5 : 309452.dasm (-0.92% of base)
          -5 : 78557.dasm (-0.47% of base)
          -4 : 125040.dasm (-0.16% of base)
          -3 : 126900.dasm (-0.13% of base)
          -1 : 81013.dasm (-0.30% of base)
          -1 : 81014.dasm (-0.28% of base)

13 total files with Code Size differences (12 improved, 1 regressed), 0 unchanged.

Top method regressions (bytes):
           4 ( 0.60% of base) : 127375.dasm - Microsoft.CodeAnalysis.VisualBasic.Simplification.Simplifiers.NameSimplifier:CanSimplifyNullable(Microsoft.CodeAnalysis.INamedTypeSymbol,Microsoft.CodeAnalysis.VisualBasic.Syntax.NameSyntax):bool

Top method improvements (bytes):
        -131 (-1.49% of base) : 126763.dasm - Microsoft.CodeAnalysis.VisualBasic.Extensions.ParenthesizedExpressionSyntaxExtensions:CanRemoveParentheses(Microsoft.CodeAnalysis.VisualBasic.Syntax.ParenthesizedExpressionSyntax,Microsoft.CodeAnalysis.SemanticModel,System.Threading.CancellationToken):bool
         -43 (-25.60% of base) : 126729.dasm - Microsoft.CodeAnalysis.VisualBasic.Extensions.ExpressionSyntaxExtensions:IsAnyLiteralExpression(Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax):bool
         -23 (-2.85% of base) : 126559.dasm - Microsoft.CodeAnalysis.VisualBasic.CodeGeneration.VisualBasicSyntaxGenerator:WithTypeArguments(Microsoft.CodeAnalysis.SyntaxNode,System.Collections.Generic.IEnumerable`1[[Microsoft.CodeAnalysis.SyntaxNode, Microsoft.CodeAnalysis, Version=3.8.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]]):Microsoft.CodeAnalysis.SyntaxNode:this
         -13 (-7.10% of base) : 126582.dasm - Microsoft.CodeAnalysis.VisualBasic.CodeGeneration.VisualBasicSyntaxGenerator:ParenthesizeLeft(Microsoft.CodeAnalysis.SyntaxNode):Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax
          -5 (-0.94% of base) : 10565.dasm - Microsoft.CodeAnalysis.SolutionState:GetProjectIdDirectly(Microsoft.CodeAnalysis.ISymbol,System.Runtime.CompilerServices.ConditionalWeakTable`2[[Microsoft.CodeAnalysis.ISymbol, Microsoft.CodeAnalysis, Version=3.8.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35],[Microsoft.CodeAnalysis.ProjectId, Microsoft.CodeAnalysis.Workspaces, Version=3.8.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]]):Microsoft.CodeAnalysis.ProjectId:this
          -5 (-0.47% of base) : 212745.dasm - System.Tests.StringTests:EqualsTest(System.String,System.Object,int,bool)
          -5 (-0.92% of base) : 309452.dasm - System.Security.Permissions.Tests.PrincipalPermissionTest:Intersect_IntersectionInRole():this
          -5 (-0.47% of base) : 78557.dasm - System.Tests.StringTests:EqualsTest(System.String,System.Object,int,bool)
          -4 (-0.16% of base) : 125040.dasm - Expander:FullyQualifyIdentifierName(Microsoft.CodeAnalysis.ISymbol,Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax,Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax,bool,bool,bool):Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax:this
          -3 (-0.13% of base) : 126900.dasm - Microsoft.CodeAnalysis.VisualBasic.Extensions.CastAnalyzer:IsUnnecessary():bool:this
          -1 (-0.30% of base) : 81013.dasm - System.Tests.StackTraceHiddenAttributeTests:ClassHidden_EnvironmentStackTrace():this
          -1 (-0.28% of base) : 81014.dasm - System.Tests.StackTraceHiddenAttributeTests:StructHidden_EnvironmentStackTrace():this

Top method regressions (percentages):
           4 ( 0.60% of base) : 127375.dasm - Microsoft.CodeAnalysis.VisualBasic.Simplification.Simplifiers.NameSimplifier:CanSimplifyNullable(Microsoft.CodeAnalysis.INamedTypeSymbol,Microsoft.CodeAnalysis.VisualBasic.Syntax.NameSyntax):bool

Top method improvements (percentages):
         -43 (-25.60% of base) : 126729.dasm - Microsoft.CodeAnalysis.VisualBasic.Extensions.ExpressionSyntaxExtensions:IsAnyLiteralExpression(Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax):bool
         -13 (-7.10% of base) : 126582.dasm - Microsoft.CodeAnalysis.VisualBasic.CodeGeneration.VisualBasicSyntaxGenerator:ParenthesizeLeft(Microsoft.CodeAnalysis.SyntaxNode):Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax
         -23 (-2.85% of base) : 126559.dasm - Microsoft.CodeAnalysis.VisualBasic.CodeGeneration.VisualBasicSyntaxGenerator:WithTypeArguments(Microsoft.CodeAnalysis.SyntaxNode,System.Collections.Generic.IEnumerable`1[[Microsoft.CodeAnalysis.SyntaxNode, Microsoft.CodeAnalysis, Version=3.8.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]]):Microsoft.CodeAnalysis.SyntaxNode:this
        -131 (-1.49% of base) : 126763.dasm - Microsoft.CodeAnalysis.VisualBasic.Extensions.ParenthesizedExpressionSyntaxExtensions:CanRemoveParentheses(Microsoft.CodeAnalysis.VisualBasic.Syntax.ParenthesizedExpressionSyntax,Microsoft.CodeAnalysis.SemanticModel,System.Threading.CancellationToken):bool
          -5 (-0.94% of base) : 10565.dasm - Microsoft.CodeAnalysis.SolutionState:GetProjectIdDirectly(Microsoft.CodeAnalysis.ISymbol,System.Runtime.CompilerServices.ConditionalWeakTable`2[[Microsoft.CodeAnalysis.ISymbol, Microsoft.CodeAnalysis, Version=3.8.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35],[Microsoft.CodeAnalysis.ProjectId, Microsoft.CodeAnalysis.Workspaces, Version=3.8.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]]):Microsoft.CodeAnalysis.ProjectId:this
          -5 (-0.92% of base) : 309452.dasm - System.Security.Permissions.Tests.PrincipalPermissionTest:Intersect_IntersectionInRole():this
          -5 (-0.47% of base) : 212745.dasm - System.Tests.StringTests:EqualsTest(System.String,System.Object,int,bool)
          -5 (-0.47% of base) : 78557.dasm - System.Tests.StringTests:EqualsTest(System.String,System.Object,int,bool)
          -1 (-0.30% of base) : 81013.dasm - System.Tests.StackTraceHiddenAttributeTests:ClassHidden_EnvironmentStackTrace():this
          -1 (-0.28% of base) : 81014.dasm - System.Tests.StackTraceHiddenAttributeTests:StructHidden_EnvironmentStackTrace():this
          -4 (-0.16% of base) : 125040.dasm - Expander:FullyQualifyIdentifierName(Microsoft.CodeAnalysis.ISymbol,Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax,Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax,bool,bool,bool):Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax:this
          -3 (-0.13% of base) : 126900.dasm - Microsoft.CodeAnalysis.VisualBasic.Extensions.CastAnalyzer:IsUnnecessary():bool:this

13 total methods with Code Size differences (12 improved, 1 regressed), 0 unchanged.

```

</details>

--------------------------------------------------------------------------------

**Comment by EgorBo** [comment]
@AndyAyersMS I wonder why my fix had a bigger diff - it found more opportunities to fold?

File: src/coreclr/jit/redundantbranchopts.cpp [hunk 1]
```diff
@@ -264,17 +264,38 @@ bool Compiler::optJumpThread(BasicBlock* const block, BasicBlock* const domBlock
...
+                return false;
+            }
+            JITDUMP(" -- bypassing %sdom " FMT_BB " as it was already optimized\n",
+                    (idomBlock == block->bbIDom) ? "i" : "", idomBlock->bbNum);
+            idomBlock = idomBlock->bbIDom;
+        }
+
+        // If we didn't bail out above, we should have reached domBlock.
+        //
+        assert(idomBlock == domBlock);
```
Comment (EgorBo) [inline src/coreclr/jit/redundantbranchopts.cpp:294]: Do we need to check that the new dominator is in the same EH region here?

Comment (AndyAyersMS) [reply src/coreclr/jit/redundantbranchopts.cpp:294]: Yes. 

EH region constraints are checked later on, when `optJumpThread` examines exactly how control leaving `domBlock` reaches the block to be optimized.

**Review by EgorBo — APPROVED** [review]

**Comment by AndyAyersMS** [comment]
> @AndyAyersMS I wonder why my fix had a bigger diff - it found more opportunities to fold?

Hard to say for sure, but typically with these opts there are dependent cases, so that optimizing in one place reveals new opportunities in another, and it's hard to get all of these cleaned up in a single pass.

**Pull request merged**